lean/main/overview.lean

```diff
@@ -0,0 +1,200 @@
+
+The behaviour of syntax sugars (`elab`, `macro`, etc.) can be understood from these first principles.
+
+## Manual conversions between `Syntax`/`Expr`/executable-code
+
+Lean will execute the afforementioned **parsing**/**elaboration**/**evaluation** steps for you automatically if you use `syntax`, `macro` and `elab` commands, however, when you're writing your tactics, you will also frequently need to perform these transitions manually. You can use the following functions for that:
+
+<p align="center">
+<img width="650px" src="https://github.com/arthurpaulino/lean4-metaprogramming-book/assets/7578559/b403e650-dab4-4843-be8c-8fb812695a3a"/>
+</p>
+
+Note how all functions that let us turn `Syntax` into `Expr` start with "elab", short for "elaboration"; and all functions that let us turn `Expr` (or `Syntax`) into `actual code` start with "eval", short for "evaluation".
+
+## Assigning meaning: macro VS elaboration?
+
+In principle, you can do with a `macro` (almost?) anything you can do with the `elab` function. Just write what you would have in the body of your `elab` as a syntax within `macro`. However, the rule of thumb here is to only use `macro`s when the conversion is simple and truly feels elementary to the point of aliasing. As Henrik Böving puts it: "as soon as types or control flow is involved a macro is probably not reasonable anymore" ([Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/The.20line.20between.20term.20elaboration.20and.20macro/near/280951290)). So - use `macro`s for creating syntax sugars, notations, and shortcuts, and prefer `elab`s for writing out code with some programming logic, even if it's potentially implementable in a `macro`.
+
+## Additional comments
+
+Finally - some notes that should clarify a few things as you read the coming chapters.
+
+### Printing Messages
+
+In the `#assertType` example, we used `logInfo` to make our command print
+something. If, instead, we just want to perform a quick debug, we can use
+`dbg_trace`.
+
+They behave a bit differently though, as we can see below:
+-/
+
+elab "traces" : tactic => do
+  let array := List.replicate 2 (List.range 3)
+  Lean.logInfo m!"logInfo: {array}"
+  dbg_trace f!"dbg_trace: {array}"
+
+example : True := by -- `example` is underlined in blue, outputting:
+                     -- dbg_trace: [[0, 1, 2], [0, 1, 2]]
+  traces -- now `traces` is underlined in blue, outputting
+         -- logInfo: [[0, 1, 2], [0, 1, 2]]
+  trivial
+
+/-
+### Type correctness
+
+Since the objects defined in the meta-level are not the ones we're most
+interested in proving theorems about, it can sometimes be overly tedious to
+prove that they are type correct. For example, we don't care about proving that
+a recursive function to traverse an expression is well founded. Thus, we can
+use the `partial` keyword if we're convinced that our function terminates. In
+the worst case scenario, our function gets stuck in a loop but the kernel is
```

Comment:
And Lean Server crashes in vscode :)

`but the kernel is not reached/affected` -> you probably wanted to say "soundness of the underlying type theory implemented in kernel isn't affected"

Reply:
https://github.com/arthurpaulino/lean4-metaprogramming-book/pull/99/commits/d52680954edc8fcff5b596ba736cfc457172f956 is this ok?